libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/ActivityChecker.cs

```diff
@@ -0,0 +1,372 @@
+                    case "type":
+                    case "text":
+                    case "speak":
+                    case "inputhint":
+                    case "attachmentlayout":
+                        break;
+
+                    case "attachments":
+                        result.AddRange(CheckAttachments(value));
+                        break;
+
+                    case "suggestedactions":
+                        result.AddRange(CheckSuggestions(value));
+                        break;
+
+                    case "attachment":
+                        result.Add(BuildDiagnostic($"'{property}' is not support, do you mean 'attachments'?", false));
+                        break;
+                    case "suggestedaction":
+                        result.Add(BuildDiagnostic($"'{property}' is not support, do you mean 'suggestedactions'?", false));
+                        break;
+                    default:
+                        result.Add(BuildDiagnostic($"'{property}' is not support in message activity.", false));
+                        break;
+                }
+            }
+
+            return result;
+        }
+
+        private static List<Diagnostic> CheckSuggestions(JToken value)
+        {
+            var result = new List<Diagnostic>();
+            var actions = NormalizedToList(value);
+            foreach (var action in actions)
+            {
+                if (action is JValue jValue && jValue.Type == JTokenType.String)
+                {
+                    return result;
+                }
+                else if (action is JObject actionJObj)
+                {
+                    result.AddRange(CheckCardAction(actionJObj));
+                }
+            }
+
+            return result;
+        }
+
+        private static List<Diagnostic> CheckButtons(JToken value)
+        {
+            var result = new List<Diagnostic>();
+            var actions = NormalizedToList(value);
+
+            foreach (var action in actions)
+            {
+                if (action is JValue jValue && jValue.Type == JTokenType.String)
+                {
+                    return result;
+                }
+                else if (action is JObject actionJObj)
+                {
+                    result.AddRange(CheckCardAction(actionJObj));
+                }
+            }
+
+            return result;
+        }
+
+        private static List<Diagnostic> CheckCardAction(JObject cardActionJObj)
+        {
+            var result = new List<Diagnostic>();
+            var type = GetStructureType(cardActionJObj);
+            if (type != nameof(CardAction).ToLower())
+            {
+                result.Add(BuildDiagnostic($"'{type}' is not card action type.", false));
+            }
+            else
+            {
+                foreach (var item in cardActionJObj)
+                {
+                    var property = item.Key.Trim();
+                    var value = item.Value.ToString().ToLower();
```

Comment:
All of these "ToLower" calls throughout the code appear to be wrong. 

These should call a single "Normalize" method, and that method should proabbaly be calling the ToLowerInvariant() method. As it sits, all of the ToLower calls are going to be culture aware and have issues on certain platforms.